### ISSUE TITLE
[`v3`] CachedGISTEmbedLoss does not work with `eval_dataset`

### ISSUE THREAD

**Issue by smerrill** [UKPLab/sentence-transformers#2667]
Hey - back again with more testing against commit 33c4a85944ed98b43f8b2c7ad53eeeceb4ea4d24 of the v3 release branch, and I'm writing in to show some weird behavior that I found. I have been using the InformationRetrievalEvaluator in my v2-based training pipelines (and my v3 ones as well) but I noticed that passing in `eval_dataset` to the trainer causes crashes. I can get this to work by commenting out the `eval_dataset` line, and then I get my train loss reported along with the IREvaluator metrics in my runs (and to wandb, etc.)

That said, should there be some documentation or safeguards in the code and also (partially just for my own edification) is it possible to use another evaluator or enable `eval_dataset` so that I can get dev loss reported along with the IR metrics?

A toy repro follows:

```
from sentence_transformers import (
    SentenceTransformerTrainer,
    SentenceTransformerTrainingArguments,
    evaluation,
    losses,
    SentenceTransformer,
)
import polars as pl
from datasets import Dataset

item_pairs = [
    ("toad", "frog"),
    ("dog", "pup"),
    ("cat", "feline"),
]

data = pl.DataFrame(item_pairs, schema=[("anchor", pl.String), ("positive", pl.String)])
data_columns = data.get_columns()

dataset = Dataset(data.to_arrow()).train_test_split(test_size=0.2)
train_dataset = dataset["train"]
eval_dataset = dataset["test"]

retrieval_evaluator = evaluation.InformationRetrievalEvaluator(
    {x: x for x in data_columns[0]},
    {y: y for y in data_columns[1]},
    {x: set(y) for x, y in item_pairs},
    show_progress_bar=True,
)

OUTPUT_DIR = "./test-training-output"

MODEL_NAME = "all-mpnet-base-v2"
GUIDE_MODEL_NAME = "all-MiniLM-L6-v2"

model = SentenceTransformer(MODEL_NAME)
guide_model = SentenceTransformer(GUIDE_MODEL_NAME)

retrieval_evaluator(model)

train_loss = losses.CachedGISTEmbedLoss(model, mini_batch_size=4, guide=guide_model)

training_args = SentenceTransformerTrainingArguments(
    output_dir=OUTPUT_DIR,
    num_train_epochs=1,
    per_device_train_batch_size=4,
    per_device_eval_batch_size=4,
    eval_strategy="epoch",
)

trainer = SentenceTransformerTrainer(
    model=model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset, # If I comment this line out, the training completes.
    loss=train_loss,
    args=training_args,
    evaluator=retrieval_evaluator,
)

trainer.train()

model.save(OUTPUT_DIR)
```

And here is the stack trace copied from Jupyter with some URLs and paths snipped:

```
RuntimeError Traceback (most recent call last)
Cell In[7], line 60
43 training_args = SentenceTransformerTrainingArguments(
44 output_dir=OUTPUT_DIR,
45 num_train_epochs=1,
(...)
48 eval_strategy="epoch",
49 )
51 trainer = SentenceTransformerTrainer(
52 model=model,
53 train_dataset=train_dataset,
(...)
57 evaluator=retrieval_evaluator,
58 )
---> 60 trainer.train()
62 model.save(OUTPUT_DIR)

File [/lib/python3.10/site-packages/transformers/trainer.py:1885], in Trainer.train(self, resume_from_checkpoint, trial, ignore_keys_for_eval, \*\*kwargs)
1883 hf_hub_utils.enable_progress_bars()
1884 else:
-> 1885 return inner_training_loop(
1886 args=args,
1887 resume_from_checkpoint=resume_from_checkpoint,
1888 trial=trial,
1889 ignore_keys_for_eval=ignore_keys_for_eval,
1890 )

File [/lib/python3.10/site-packages/transformers/trainer.py:2311], in Trainer.\_inner_training_loop(self, batch_size, args, resume_from_checkpoint, trial, ignore_keys_for_eval)
2308 self.control.should_training_stop = True
2310 self.control = self.callback_handler.on_epoch_end(args, self.state, self.control)
-> 2311 self.\_maybe_log_save_evaluate(tr_loss, grad_norm, model, trial, epoch, ignore_keys_for_eval)
2313 if DebugOption.TPU_METRICS_DEBUG in self.args.debug:
2314 if is_torch_xla_available():
2315 # tpu-comment: Logging debug metrics for PyTorch[/XLA](https://steven-ml-trainer.bobcat-byzantine.ts.net:8888/XLA) (compile, execute times, ops, etc.)

File [/lib/python3.10/site-packages/transformers/trainer.py:2721], in Trainer.\_maybe_log_save_evaluate(self, tr_loss, grad_norm, model, trial, epoch, ignore_keys_for_eval)
2719 metrics = None
2720 if self.control.should_evaluate:
-> 2721 metrics = self.evaluate(ignore_keys=ignore_keys_for_eval)
2722 self.\_report_to_hp_search(trial, self.state.global_step, metrics)
2724 # Run delayed LR scheduler now that metrics are populated

File [/src/sentence-transformers/sentence_transformers/trainer.py:279], in SentenceTransformerTrainer.evaluate(self, eval_dataset, ignore_keys, metric_key_prefix)
277 if isinstance(eval_dataset, DatasetDict) and isinstance(self.loss, dict):
278 eval_dataset = self.add_dataset_name_column(eval_dataset)
--> 279 return super().evaluate(eval_dataset, ignore_keys, metric_key_prefix)

File [/lib/python3.10/site-packages/transformers/trainer.py:3572], in Trainer.evaluate(self, eval_dataset, ignore_keys, metric_key_prefix)
3569 start_time = time.time()
3571 eval_loop = self.prediction_loop if self.args.use_legacy_prediction_loop else self.evaluation_loop
-> 3572 output = eval_loop(
3573 eval_dataloader,
3574 description="Evaluation",
3575 # No point gathering the predictions if there are no metrics, otherwise we defer to
3576 # self.args.prediction_loss_only
3577 prediction_loss_only=True if self.compute_metrics is None else None,
3578 ignore_keys=ignore_keys,
3579 metric_key_prefix=metric_key_prefix,
3580 )
3582 total_batch_size = self.args.eval_batch_size \* self.args.world_size
3583 if f"{metric_key_prefix}\_jit_compilation_time" in output.metrics:

File [/src/sentence-transformers/sentence_transformers/trainer.py:289], in SentenceTransformerTrainer.evaluation_loop(self, dataloader, description, prediction_loss_only, ignore_keys, metric_key_prefix)
281 def evaluation_loop(
282 self,
283 dataloader: DataLoader,
(...)
287 metric_key_prefix: str = "eval",
288 ) -> EvalLoopOutput:
--> 289 output = super().evaluation_loop(
290 dataloader=dataloader,
291 description=description,
292 prediction_loss_only=prediction_loss_only,
293 ignore_keys=ignore_keys,
294 metric_key_prefix=metric_key_prefix,
295 )
297 # If the evaluator is not defined, we can just return the output
298 if self.evaluator is None:

File [/lib/python3.10/site-packages/transformers/trainer.py:3757], in Trainer.evaluation_loop(self, dataloader, description, prediction_loss_only, ignore_keys, metric_key_prefix)
3754 batch_size = observed_batch_size
3756 # Prediction step
-> 3757 loss, logits, labels = self.prediction_step(model, inputs, prediction_loss_only, ignore_keys=ignore_keys)
3758 main_input_name = getattr(self.model, "main_input_name", "input_ids")
3759 inputs_decode = self.\_prepare_input(inputs[main_input_name]) if args.include_inputs_for_metrics else None

File [/lib/python3.10/site-packages/transformers/trainer.py:3971], in Trainer.prediction_step(self, model, inputs, prediction_loss_only, ignore_keys)
3969 if has_labels or loss_without_labels:
3970 with self.compute_loss_context_manager():
-> 3971 loss, outputs = self.compute_loss(model, inputs, return_outputs=True)
3972 loss = loss.mean().detach()
3974 if isinstance(outputs, dict):

File [/src/sentence-transformers/sentence_transformers/trainer.py:226], in SentenceTransformerTrainer.compute_loss(self, model, inputs, return_outputs)
220 if (
221 self.args.parallel_mode != ParallelMode.NOT_PARALLEL
222 and hasattr(model, "module")
223 and hasattr(loss_fn, "model")
224 ):
225 loss_fn = self.override_model_in_loss(loss_fn, model)
--> 226 loss = loss_fn(features, labels)
227 if return_outputs:
228 # During prediction[/evaluation](https://steven-ml-trainer.bobcat-byzantine.ts.net:8888/evaluation), `compute_loss` will be called with `return_outputs=True`.
229 # However, Sentence Transformer losses do not return outputs, so we return an empty dictionary.
230 # This does not result in any problems, as the SentenceTransformersTrainingArguments sets
231 # `prediction_loss_only=True` which means that the output is not used.
232 return loss, {}

File [/lib/python3.10/site-packages/torch/nn/modules/module.py:1511], in Module.\_wrapped_call_impl(self, *args, \*\*kwargs)
1509 return self.\_compiled_call_impl(*args, **kwargs) # type: ignore[misc]
1510 else:
-> 1511 return self.\_call_impl(\*args, **kwargs)

File [/lib/python3.10/site-packages/torch/nn/modules/module.py:1520], in Module.\_call_impl(self, *args, \*\*kwargs)
1515 # If we don't have any hooks, we want to skip the rest of the logic in
1516 # this function, and just call forward.
1517 if not (self.\_backward_hooks or self.\_backward_pre_hooks or self.\_forward_hooks or self.\_forward_pre_hooks
1518 or \_global_backward_pre_hooks or \_global_backward_hooks
1519 or \_global_forward_hooks or \_global_forward_pre_hooks):
-> 1520 return forward_call(*args, \*\*kwargs)
1522 try:
1523 result = None

File [/src/sentence-transformers/sentence_transformers/losses/CachedGISTEmbedLoss.py:302], in CachedGISTEmbedLoss.forward(self, sentence_features, labels)
299 self.random_states.append(random_state_mbs)
301 # Step (2): Calculate the loss, backward up to the embeddings and cache the gradients wrt. to the embeddings
--> 302 loss = self.calculate_loss_and_cache_gradients(reps, reps_guided)
303 # Step (3): A 2nd embedding step with gradients[/computation](https://steven-ml-trainer.bobcat-byzantine.ts.net:8888/computation) graphs and connect the cached gradients into the backward chain
304 loss.register_hook(partial(\_backward_hook, sentence_features=sentence_features, loss_obj=self))

File [/src/sentence-transformers/sentence_transformers/losses/CachedGISTEmbedLoss.py:271], in CachedGISTEmbedLoss.calculate*loss_and_cache_gradients(self, reps, reps_guided)
269 scores = scores [/](https://steven-ml-trainer.bobcat-byzantine.ts.net:8888/) self.temperature
270 loss_mbatch: torch.Tensor = self.cross_entropy_loss(scores, labels[b:e]) \* len(scores) [/](https://steven-ml-trainer.bobcat-byzantine.ts.net:8888/) batch_size
--> 271 loss_mbatch.backward()
272 losses.append(loss_mbatch.detach())
274 loss = sum(losses).requires_grad*()

File [/lib/python3.10/site-packages/torch/\_tensor.py:522], in Tensor.backward(self, gradient, retain_graph, create_graph, inputs)
512 if has_torch_function_unary(self):
513 return handle_torch_function(
514 Tensor.backward,
515 (self,),
(...)
520 inputs=inputs,
521 )
--> 522 torch.autograd.backward(
523 self, gradient, retain_graph, create_graph, inputs=inputs
524 )

File [/lib/python3.10/site-packages/torch/autograd/**init**.py:266], in backward(tensors, grad*tensors, retain_graph, create_graph, grad_variables, inputs)
261 retain_graph = create_graph
263 # The reason we repeat the same comment below is that
264 # some Python versions print out the first line of a multi-line function
265 # calls in the traceback and some print out the last line
--> 266 Variable.\_execution_engine.run_backward( # Calls into the C++ engine to run the backward pass
267 tensors,
268 grad_tensors*,
269 retain_graph,
270 create_graph,
271 inputs,
272 allow_unreachable=True,
273 accumulate_grad=True,
274 )

RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn
```

**Comment by tomaarsen** [comment]
Hello!

Big thanks for reporting this. I think I was able to chase this down and fix it in #2668. The toy reproduction code now works for me, but if you can, please check whether your "real" code also works with that branch. This should do it:
```
pip install -U git+https://github.com/tomaarsen/sentence-transformers.git@v3/fix_cached_eval
```

- Tom Aarsen

**Comment by tomaarsen** [comment]
Closing this as #2668 was merged into the v3.0-pre-release branch. It'll be included in next week's release.

- Tom Aarsen
